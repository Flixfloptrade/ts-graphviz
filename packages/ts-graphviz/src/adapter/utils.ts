export { createCommandAndArgs } from './utils/create-command-and-args.js';
