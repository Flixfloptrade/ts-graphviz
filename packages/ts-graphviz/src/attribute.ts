import type { Attribute } from '@ts-graphviz/common';

/**
 * Attribute constant.
 * @public
 */
export const attribute: Attribute.keys = new Proxy(
  Object.freeze({}) as Attribute.keys,
  {
    get: (_, key: string) => key,
  },
);
