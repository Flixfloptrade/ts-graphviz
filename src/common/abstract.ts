import { IDot } from '../common';
// tslint:disable: max-classes-per-file

/**
 * @category Base
 */
export abstract class GraphvizObject {}

/**
 * @category Base
 */
export abstract class DotBase extends GraphvizObject implements IDot {
  public abstract toDot(): string;
}
