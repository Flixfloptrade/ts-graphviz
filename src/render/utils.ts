export function escape(str: string): string {
  return str.replace(/\n/g, '\\n').replace(/"/g, '\\"');
}

export function wrap(word: string, wrapper: string): string {
  return `${wrapper}${word}${wrapper}`;
}

export function wrapFactory(wrapper: string): (word: string) => string {
  return (word: string): string => wrap(word, wrapper);
}

export function leftPad(word: string, pad: string): string {
  return `${pad}${word}`;
}

export function leftPadFactory(pad: string): (w?: string) => string | undefined {
  return (w?: string): string | undefined => {
    if (typeof w === 'string') {
      return leftPad(w, pad);
    }
  };
}

export const quote = wrapFactory('"');

export const spaceLeftPad = leftPadFactory(' ');

export function concatWordsFactory(deciliter: string): (...lines: (string | undefined)[]) => string {
  return (...lines: (string | undefined)[]): string => lines.filter((l) => typeof l === 'string').join(deciliter);
}

export const concatWordsWithSpace = concatWordsFactory(' ');

export const concatWordsWithColon = concatWordsFactory(':');

export const joinLines = concatWordsFactory('\n');
export const join = concatWordsFactory('');
export function joinWith(separator: string, lines: (string | undefined)[]): string {
  return lines.filter((l) => typeof l === 'string').join(separator);
}

export function indent(src: string): string {
  const space = '  ';
  return src
    .split('\n')
    .map((l) => join(space, l))
    .join('\n');
}

export function commentOut(src: string): string {
  return src
    .split('\n')
    .map((l) => join('// ', l).trim())
    .join('\n');
}
export function commentOutIfExist(src: string | undefined): string | undefined {
  return typeof src === 'string' ? commentOut(src) : undefined;
}
