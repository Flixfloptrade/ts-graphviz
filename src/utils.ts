export * from './utils/type.js';
export * from './utils/functional.js';
