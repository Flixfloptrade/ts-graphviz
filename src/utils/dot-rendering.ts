/**
 * @hidden
 */
export function escape(str: string): string {
  return str.replace(/\n/g, '\\n').replace(/"/g, '\\"');
}

/**
 * @hidden
 */
export function quote(src: string): string {
  return `"${src}"`;
}

/**
 * @hidden
 */
export function concatWords(...lines: (string | undefined)[]): string {
  return lines.filter(l => typeof l === 'string').join(' ');
}

/**
 * @hidden
 */
export function joinLines(...lines: (string | undefined)[]): string {
  return lines.filter(l => typeof l === 'string').join('\n');
}

/**
 * @hidden
 */
export function indent(src: string, depth: number = 1): string {
  const space = ' '.repeat(2 * depth);
  return src
    .split('\n')
    .map(l => `${space}${l}`)
    .join('\n');
}
