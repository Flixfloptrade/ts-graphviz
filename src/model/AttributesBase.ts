import { DotBase } from '../abstract';
import { AttributesObject, AttributesValue, IAttributesBase } from '../types';

/**
 * @hidden
 */
export abstract class AttributesBase<T extends string> extends DotBase implements IAttributesBase<T> {
  /** @hidden */
  protected attrs: Map<T, AttributesValue> = new Map();

  public entries(): IterableIterator<[T, AttributesValue]> {
    return this.attrs.entries();
  }

  /** The size of the attribute. */
  get size(): number {
    return this.attrs.size;
  }
  /** The size of the attribute. */
  public get(key: T): AttributesValue | undefined {
    return this.attrs.get(key);
  }
  /** Set a value to the attribute. */
  public set(key: T, value: AttributesValue): void {
    this.attrs.set(key, value);
  }

  public delete(key: T): void {
    this.attrs.delete(key);
  }

  public apply(attributes: AttributesObject<T>): void {
    for (const [key, value] of Object.entries(attributes)) {
      this.set(key as T, value as string | boolean | number);
    }
  }

  public clear(): void {
    this.attrs.clear();
  }
}
