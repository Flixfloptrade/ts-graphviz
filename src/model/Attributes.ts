import { DotBase } from '../common';
import { indent } from '../utils/dot-rendering';
import { Literal } from './Literal';

/**
 * @category Attributes
 */
export class Attributes extends DotBase {
  protected attrs: Map<string, Literal> = new Map();
  get size(): number {
    return this.attrs.size;
  }
  public get(key: string): Literal | undefined {
    return this.attrs.get(key);
  }
  public set(key: string, value: any): void {
    if (value instanceof Literal) {
      this.attrs.set(key, value);
    } else {
      this.attrs.set(key, new Literal(value));
    }
  }
  public toDot(): string {
    if (this.size === 0) {
      return '';
    }
    return [
      '[',
      ...Array.from(this.attrs.entries()).map(([key, value]) => indent(`${key} = ${value.toDot()},`)),
      ']',
    ].join('\n');
  }
}
