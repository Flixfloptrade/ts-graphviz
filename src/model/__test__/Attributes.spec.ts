import { DotBase, GraphvizObject } from '../../abstract';
import { attribute } from '../../attribute';
import { Attributes } from '../Attributes';
import { AttributesBase } from '../AttributesBase';
import { toDot } from '../../utils/dot-rendering';

describe('class Attributes', () => {
  let attrs: Attributes<attribute.Attribute>;
  beforeEach(() => {
    attrs = new Attributes();
  });

  it('should be instance of Attributes/AttributesBase/DotBase/GraphvizObject', () => {
    expect(attrs).toBeInstanceOf(Attributes);
    expect(attrs).toBeInstanceOf(AttributesBase);
    expect(attrs).toBeInstanceOf(DotBase);
    expect(attrs).toBeInstanceOf(GraphvizObject);
  });

  it('size should be 0 by default', () => {
    expect(attrs.size).toBe(0);
  });

  describe('renders correctly by toDot method', () => {
    it('no attributes', () => {
      expect(toDot(attrs)).toMatchSnapshot();
    });

    it('one attribute', () => {
      attrs.set('label', 'test');
      expect(toDot(attrs)).toMatchSnapshot();
    });

    it('some attributes', () => {
      attrs.set('label', 'test');
      attrs.set('color', 'red');
      expect(toDot(attrs)).toMatchSnapshot();
    });

    test('set some attributes by apply', () => {
      attrs.apply({
        label: 'this is test',
        color: 'red',
        fontsize: 16,
      });
      expect(toDot(attrs)).toMatchSnapshot();
    });

    test('apply/clear attribute', () => {
      attrs.apply({
        label: 'this is test',
        color: 'red',
        fontsize: 16,
      });
      expect(attrs.size).toBe(3);
      attrs.clear();
      expect(attrs.size).toBe(0);
    });

    test('set/get/delete attribute', () => {
      const id = 'test';
      attrs.set('label', id);
      expect(attrs.get('label')?.value).toBe(id);
      attrs.delete('label');
      expect(attrs.get('label')).toBeUndefined();
    });

    describe('edge with comment', () => {
      beforeEach(() => {
        attrs.set('label', 'test');
        attrs.set('color', 'red');
      });
      test('single line comment', () => {
        attrs.comment = 'this is comment.';
        expect(toDot(attrs)).toMatchSnapshot();
      });

      test('multi line comment', () => {
        attrs.comment = 'this is comment.\nsecond line.';
        expect(toDot(attrs)).toMatchSnapshot();
      });
    });
  });
});
