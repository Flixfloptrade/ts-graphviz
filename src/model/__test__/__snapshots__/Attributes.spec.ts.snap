// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class Attributes renders correctly by toDot method no attributes 1`] = `""`;

exports[`class Attributes renders correctly by toDot method one attribute 1`] = `
"[
  label = test,
]"
`;

exports[`class Attributes renders correctly by toDot method some attributes 1`] = `
"[
  label = test,
  color = red,
]"
`;
